mempool/iterators.go

```diff
@@ -0,0 +1,208 @@
+package mempool
+
+import (
+	"fmt"
+
+	"github.com/cometbft/cometbft/internal/clist"
+	"github.com/cometbft/cometbft/types"
+)
+
+// WRRIterator is the base struct for implementing iterators that traverse lanes with
+// the classical Weighted Round Robin (WRR) algorithm.
+type WRRIterator struct {
+	sortedLanes []types.Lane
+	laneIndex   int                            // current lane being iterated; index on sortedLanes
+	counters    map[types.Lane]uint            // counters of consumed entries, for WRR algorithm
+	cursors     map[types.Lane]*clist.CElement // last accessed entries on each lane
+}
+
+func (iter *WRRIterator) nextLane() types.Lane {
+	iter.laneIndex = (iter.laneIndex + 1) % len(iter.sortedLanes)
+	return iter.sortedLanes[iter.laneIndex]
+}
+
+// Non-blocking version of the WRR iterator to be used for reaping and
+// rechecking transactions.
+//
+// Lock must be held on the mempool when iterating: the mempool cannot be
+// modified while iterating.
+type NonBlockingWRRIterator struct {
+	WRRIterator
+}
+
+func NewWRRIterator(mem *CListMempool) *NonBlockingWRRIterator {
```

Comment:
Name looks confusing to me. Can we rename to `NewNonBlockingWRRIterator`? This way there's no room for ambiguities.

Reply:
Maybe to shorten all names, we could do:
* `NewNonBlockingIterator`
* `NewBlockingIterator`
(just a thought)

Reply:
In #4013 I use the names `BlockingIterator` and `NonBlockingIterator` for two new interfaces, and I keep `BlockingWRRIterator` and `NonBlockingWRRIterator` for the implementations.